src/main/menu.js

```diff
@@ -0,0 +1,420 @@
+    },
+    exit: {
+        label: "{that}.options.menuLabels.exit",
+        click: "onExit"
+    }
+});
+
+
+/*
+ ** Component to generate the menu tree structure that is relayed to gpii.app for display.
+ */
+fluid.defaults("gpii.app.menu", {
+    gradeNames: "fluid.modelComponent",
+    model: {
+        // Expected as configuration
+        //keyedInUserToken: null,
+        //snapsetName: null,
+        //preferences: {
+        //    sets: null,
+        //    activeSet: null
+        //},
+        // locally updated
+        preferenceSetsMenuItems: [],  // Updated on `preferences` changed.
+        keyedInSnapset: null,        // Must be updated when keyedInUserToken changes.
+        keyOut: null,                 // May or may not be in the menu, must be updated when keyedInUserToken changes.
+        menuTemplate: []              // This is updated on change of keyedInUserToken.
+    },
+    modelRelay: {
+        "keyedInSnapset": {
+            target: "keyedInSnapset",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                func: "gpii.app.menu.getKeyedInSnapset",
+                args: ["{that}.model.keyedInUserToken", "{that}.model.snapsetName", "{that}.options.menuLabels.keyedIn"]
+            }
+        },
+        "keyOut": {
+            target: "keyOut",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                func: "gpii.app.menu.getKeyOut",
+                args: ["{that}.model.keyedInUserToken", "{that}.options.menuLabels.keyOut", "{that}.options.menuLabels.notKeyedIn"]
+            }
+        },
+        "showPSP": {
+            target: "showPSP",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                func: "gpii.app.menu.getShowPSP",
+                args: ["{that}.model.keyedInUserToken", "{that}.options.menuLabels.psp"]
+            }
+        },
+        "preferenceSetsMenuItems": {
+            target: "preferenceSetsMenuItems",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                func: "gpii.app.menu.getPreferenceSetsMenuItems",
+                args: ["{that}.model.preferences.sets", "{that}.model.preferences.activeSet"]
+            }
+        },
+        "menuTemplate": {
+            target: "menuTemplate",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                func: "gpii.app.menu.generateMenuTemplate",
+                args: ["{that}.model.showPSP", "{that}.model.keyedInSnapset", "{that}.options.snapsets", "{that}.model.preferenceSetsMenuItems", "{that}.model.keyOut", "{that}.options.exit"]
+            },
+            priority: "last"
+        }
+    },
+    menuLabels: {
+        psp: "Open PSP",
+        keyedIn: "Keyed in with %snapsetName",    // string template
+        keyOut: "Key-out of GPII",
+        notKeyedIn: "(No one keyed in)",
+
+        // XXX DEV items
+        exit: "Exit GPII",
```

Comment:
We should factor these dev items into a separate grade so it is easier to produce production builds of the app